Making callbacks in OTP behaviours optional

zhird opened the pull request:
When implementing e.g. a gen_server, one doesn't always need all the callbacks, but end up copy
pasting or writing "empty" functions since they has to be exported to fit the behaviour. This
pr makes callbacks that are not always needed optional for the otp behaviours.

The question now is if we have chosen the right defaults for the callbacks?
I list the defaults below, feedback is welcome!

gen_event:

````erlang
handle_call(Request, State) ->
    % defaults to a crash
handle_info(_Info, State) ->
    {ok, State}. % just drops the info message
terminate(_Arg, _State) ->
    ok. % does no cleanup
code_change(_OldVsn, State, _Extra) ->
    {ok, State}. % assumes that no state change is needed
````

gen_fsm:
````erlang
handle_event(Event, StateName, StateData) ->
    % defaults to a crash
handle_sync_event(Event, From, StateName, StateData) ->
    % defaults to a crash
handle_info(_Info, StateName, StateData) ->
    {next_state, StateName, StateData}. % just drops the info message
terminate(_Reason, _StateName, _StateData) ->
    ok. % does no cleanup
code_change(_OldVsn, StateName, StateData, _Extra) ->
    {ok, StateName, StateData}. % assumes that no state change is needed
````
gen_server:
````erlang
handle_call(Request, From, State) ->
    % defaults to a crash
handle_cast(Request, State) ->
    % defaults to a crash
handle_info(_Info, State) ->
    {noreply, State}. % just drops the info message
terminate(_Reason, _State) ->
    ok. % does no cleanup
code_change(_OldVsn, State, _Extra) ->
    {ok, State}. % assumes that no state change is needed
````
gen_statem:
````erlang
terminate(_Reason, _State, _Data) ->
    ok. % does no cleanup
code_change(_OldVsn, State, Data, _Extra) ->
    {ok, State, Data}. % assumes that no state change is needed
````

bullno1 commented:
+1 from me.
A lot of time, I just put in a boilerplate to make those functions crash using something like `{stop, unexpected}` anyway.

psyeugenic commented:
@bullno1 I think sending things to `error_logger` should be an opt-in instead of opt-out, i.e. not default. If you want `error_logger` messages, just add a callback for it.

essen commented:
I would +1 for terminate and code_change (very rarely used), and -1 all other callbacks.

Actually I would -10 handle_call as it makes even less sense to allow both a gen_server:call and default handle_call, it would just make things confusing.

Either way, before making that kind of decision, it would probably a good idea to have stats across many projects on how many functions are actually empty. An empty gen_server:handle_call would most likely be rare, for example. Then decide which ones are worth making optional.

fishcakez commented:
In Elixir we are considering a similar idea for `gen_server`/`GenServer`: https://github.com/elixir-lang/elixir/pull/5676. We have similar motivations plus some other issues than aren't relevant in Erlang. However the default behaviour we are currently planning is different and I thought you might be interested how and why. I am not trying to derail this about Elixir rather give the reasons why different behaviour might be desired. Of course a good rebuttal may persuade Elixir otherwise but I hope we can keep Elixir out of the conversation (if you want to comment on it please do so in Elixir PR).

Instead of calling `erlang:function_exported/3` we add an extra check on the the `undef` error handling for logging. If the module is loaded we check the top entry of the stacktrace to investigate if the entry is the callback we were calling. If it is then the reason is changed to `'callback not exported'` instead of `'function not exported'`.

We use this handling for all callbacks (except `init/1` and `format_status/2` though we get equivalent behaviour as if they were because no logging). This means we have some differences:

`handle_info/2` defaults to a crash. We do this because there is a difference between no messages excepted and messages expected and adding a catch all clause. If there are messages expected then `handle_info` is defined and its up to user if they want a catch all or not (it may or may not be an error). However if `handle_info` is not defined then no messages are expected and so it is much likely that an error has occurred. I think it feels more natural to crash on the unexpected in OTP. Note also that if there is a default implementation and a timeout (or `hibernate`) was returned from the last callback then the timeout (or hibernation) is silently cancelled. I'm not sure on this issue because a timeout isn't useful without a `handle_info` callback. However in this case the message could well be `timeout` and the bug is hidden. Perhaps `handle_info/3` was defined by mistake. Whatever has happened it seems likely there is a bug!

`code_change/3` defaults to a crash. We do this because if a code change has not been defined but an instruction has been added to an appup to do an advanced update then the state won't be changed - but the implication is that it should be or the soft update should have been used. Again whatever happened it seems likely there is a bug!

Currently we also crash on `terminate/2` in the branch for consistency but that may not be ideal behaviour - note that the real terminate reason can appear in the stacktrace arguments.

Note also that some `gen_server`s using `gen_server:enter_loop`  probably don't use `init/1`. Currently it is required to add an implementation that returns `{stop, enotsup}` or similar. Should it be optional in the behaviour info but required in the code?

fishcakez commented:
I forgot to mention, I really like the `handle_call/3` change because many people make the mistake of doing `{reply, Error, State}` - I think even some generators default to it. This can have undefined behaviour in the client process because the client may have bad state and `{error, Reason}` might be a valid reply! If a call has arrived that can't be handled the `gen_server` should crash because it is not handling its expected API. Of course it could be a bug in the client process's code/state and hopefully a crash can, at least temporarily, resolve this. But not crashing may silently hide this bug!

essen commented:
About handle_info: I would hate to go on a production node, send a message to the wrong pid, only to have it crash for no good reasons. If handle_info has to have a default, it should not be to crash, that's too dangerous.

Another issue I can see is if your gen_server calls another process and that other process happens to send a message, or have a monitor that isn't cleaned up, you probably don't want to crash either. Sometimes it might be something you want to fix (monitor), sometimes you really just don't care about the message.

I have no idea why you would want to crash on terminate. The terminate callback is almost never needed, and there's no good way to figure out if a process really needs it or not, other than it being implemented.

fishcakez commented:
The send message to wrong pid is a red herring. `handle_info` is nearly always called when the message is "caused" by the callback modules themselves. If another process is interacting with it then it will be via call and cast etc, and increasing these have the same message format with `gen_statem` using the same message protocol as `gen_server`, so it will hit the message API and cause a crash.

The unhandled monitor is also a red herring because if there are monitors then the `handle_info` clause is going to be implemented in the callback module, so the "default" implementation does not kick in.

The default in OTP is to be explicit when handling errors and these situations can be explicitly handled by implementing the callbacks.

Sorry I should have been clearer I was mentioning the Elixir PR has an issue with the terminate/2 implementation only for completeness. Lets not talk about that here as its only relevant in Elixir.

essen commented:
Your answer makes me think you view Erlang as a perfect world where nothing should happen differently than intended. That's simply not how it is.

I don't really see why a process interacting would necessarily be via call or cast. That's certainly not true. For example if you use Gun, you can go synchronous or asynchronous, and if you go synchronous you can skip some functions which results in stray messages being ultimately ignored by handle_info.

Similarly, if you have a module interface to a process that allows you to either get_next_event (async) or cancel, you often may receive stray messages after you cancel, because you can't always wait to make sure no further messages will arrive.

Monitors, to give you an example, a few years ago I fixed a bug in OTP where *gen:call* was not cleaning up the monitor message after a crash. This means that all the `catch gen:call` calls resulted in a stray message. Oops!

Stray messages in general are fairly likely to happen even if the system as a whole behaves properly. There was a stray message in SSL for a while (related to accepting connections IIRC) to give you another example.

I don't think we want to force users to handle every stray messages.

fishcakez commented:
I'm saying that if you want to handle the errors it's clearer to write the
explicit code rather than silently handle it. When working on the code you
want bugs to show up before it matters.

On 17 Feb 2017 5:17 pm, "Loïc Hoguin" <notifications@github.com> wrote:

> Your answer makes me think you view Erlang as a perfect world where
> nothing should happen differently than intended. That's simply not how it
> is.
>
> I don't really see why a process interacting would necessarily be via call
> or cast. That's certainly not true. For example if you use Gun, you can go
> synchronous or asynchronous, and if you go synchronous you can skip some
> functions which results in stray messages being ultimately ignored by
> handle_info.
>
> Similarly, if you have a module interface to a process that allows you to
> either get_next_event (async) or cancel, you often may receive stray
> messages after you cancel, because you can't always wait to make sure no
> further messages will arrive.
>
> Monitors, to give you an example, a few years ago I fixed a bug in OTP
> where *gen:call* was not cleaning up the monitor message after a crash.
> This means that all the catch gen:call calls resulted in a stray message.
> Oops!
>
> Stray messages in general are fairly likely to happen even if the system
> as a whole behaves properly. There was a stray message in SSL for a while
> (related to accepting connections IIRC) to give you another example.
>
> I don't think we want to force users to handle every stray messages.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/erlang/otp/pull/1346#issuecomment-280710690>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB6JTaki7BmoCvgXVZKEkelJFN8KXOMZks5rddYZgaJpZM4MD6e6>
> .
>

essen commented:
Note that I do not agree with silently handling messages, I am against making the handle_* callbacks optional to begin with.

fishcakez commented:
Ah, I thought your reply (after my initial comment) was promoting the
default handle_info in the PR that does.

On 17 Feb 2017 5:37 pm, "Loïc Hoguin" <notifications@github.com> wrote:

> Note that I do not agree with silently handling messages, I am against
> making the handle_* callbacks optional to begin with.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/erlang/otp/pull/1346#issuecomment-280715935>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB6JTd1AugQHRYNEMzqo0ni9KrX3k_pcks5rddq-gaJpZM4MD6e6>
> .
>

fishcakez commented:
@essen by "I send" I assumed you meant literally you, a person, interacting
with the system via the shell. Whereas all the examples you gave are what I
meant by generated by code in the process. I think they are examples where
handle info would always need to be implemented and don't fit with a
default behaviour.

Some of the examples involving catching crashes and continuing the loop.
This is rare and would likely always require a handle_info as it would be
expected to handle dirty timeouts (using gen_statem naming) in calls. Or
the process would have extra fault handling because it would be precious in
some way, then it would require adding a handle_info clause as precaution.

The other examples involve cases where handle_info would have to be
implemented to handle the expected stray messages because messages are
arriving asynchronously.

On 17 Feb 2017 5:40 pm, "James Fish" <james@fishcakez.com> wrote:

> Ah, I thought your reply (after my initial comment) was promoting the
> default handle_info in the PR that does.
>
> On 17 Feb 2017 5:37 pm, "Loïc Hoguin" <notifications@github.com> wrote:
>
>> Note that I do not agree with silently handling messages, I am against
>> making the handle_* callbacks optional to begin with.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/erlang/otp/pull/1346#issuecomment-280715935>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AB6JTd1AugQHRYNEMzqo0ni9KrX3k_pcks5rddq-gaJpZM4MD6e6>
>> .
>>
>

essen commented:
My point is that you *don't always know* there will be stray messages -especially if they are rarely caused by a race condition, as in the get_next_event/cancel example- (or in my first example, yes you can make an honest mistake on a production node while doing something manually) which means that if the default is to crash, you would always implement it, just to make sure you won't unnecessarily crash.

Also worth noting that if your program is working despite the ignored stray messages, then those stray messages will most likely not be your priority to fix, regardless of the development stage. They're a better candidate for bugs that you log and fix *when you have time* rather than be forced to fix *immediately* because your forced crash makes the program fail to work.

Add this to yet another reason I have of not wanting handle_info optional. Those messages should be consciously ignored. If handle_info is made optional, then we will have to teach users that they need to implement it anyway because they should log stray messages. We'd have to give advice similar to "I know it's optional but always implement it and log stray messages", which is not a very good case for making it optional to begin with.

IngelaAndin commented:
Ï think handle_call and handle_cast should default to crash as suggested. I also think it is normal to crash on unexpected  calls or casts as that would be a programming error or somebody violation the API. 

I would say it is very uncommon to have a process that does not have either handle_call or handle_cast
but I know of at least one example, the ssh_daemon_channel behaviour. Also it is not evident that
you always need both calls and casts. 

I also agree that in many cases you want handle_info to throw away unexpected messages for robustness reasons, an info message may more easily be sent to a process by "accident" . I am not sure it is a good idea to default to throwing away messages. It might be confusing to the user, and it might depend on if you are implementing a server or a client process if it makes sense.

fishcakez commented:
> If handle_info is made optional, then we will have to teach users that they need to implement it anyway because they should log stray messages.

If the advise is to always implement logging because stray message should be log, wouldn't it be more convenient to have an optional callback that logs the message if its not implemented, as @IngelaAndin suggested?

IngelaAndin commented:
You do not always want to log stray message it is possible this could be a used as a DOS attack. I do not think there is a general enough default for handle_info, and hence maybe it is not a good idea to make it optional, it can be a good thing to force the user to think of what should happen.

RaimoNiskanen commented:
I agree with @fishcakez and @essen on most points here.

The callbacks `handle_call/2,3`, `handle_cast/2`, `handle_info/2,3`, `handle_event/3` and `handle_sync_event/4` I think all should in runtime crash the server if they are not implemented.  This could be done as today by just calling them, or by `try...catch`:ing and refine the error reason from `undef` to `undef_callback` as it is done in `gen_statem` today and as @fishcakez explain was considered for Elixir.

I also think it is best that the above are declared as mandatory in the behaviour definition since they have no good defaults.  Possible defaults are at least; ignore, reply `ok` (for `handle_call/2,3`), log (and ignore), and crash the server.  It depends very much on what you are writing and in which project to decide what is a good default (even how to log has to be decided), so I think it should be a forced conscious decision hence mandatory callback and default server crash.

The function `init/1` is technically optional thanks to the `GenSomething:enter_loop(...)` functions, so it is really annoying that it is declared mandatory in the behaviour definitions.  It should be optional, but crash the server if not implemented (again, look at `gen_statem`).  Possibly with a refined error reason as mentioned above.

Regarding `terminate/2,3`: it has a good default and that is to do nothing, so it is a good idea to have it declared as optional in the behaviour and runtime check with `erlang:function_exported/3` with a default implementation `-> ok`.

Function `code_change/3,4`, like @fishcakez says; it is deliberately invoked through `{advanced,Extra}` in an [`appup`](http://erlang.org/doc/man/appup.html) `upgrade` instruction because you need to rewrite your state data, so it should crash the server if not implemented.  But I see no compelling reason why it should be mandatory in the behaviour declaration.  If you write code for soft upgrade then you can add the `code_change/3,4` function and you *will* test your upgrade.  I guess that many enough projects do not use soft upgrades so having to implement a stub is more annoying than useful.

bjorng commented:
I have a probably controversial opinion about the `gen_server` callbacks `handle_call()` and `handle_cast()`.

It is not about the run-time behavior. I think they should crash if not implemented (I think we all can agree about that).

I think that both `handle_call()` and `handle_cast()` should be optional. 

Why?

Because usually you only really need to implement one of them. Having to implement the other one is just annoying. The default of crashing for the one you don't implement is a good default. Stray messages are unlikely to be interpreted as either `handle_call()` or `handle_cast()`.

zhird commented:
Great discussion here! :) 

I agree with @bjorng and really think `handle_call` and `handle_cast` should be optional, and for the same reason also `handle_event` and `handle_sync_event`. But I'll change the error reason for the crashes to `undef_callback` since that is a lot clearer.

`handle_info` is trickier though since it doesn't really have a good default implementation. The correct one would be to crash it, but unfortunately that's not very useful in reality. I think I'll remove it from the pr for now.

Regarding `init`; I didn't think about `enter_loop`... It's a good idea to make it optional as well, so I'll add that.

And I agree with @fishcakez and @RaimoNiskanen about code_change; it should crash really, so I'll update that too.

RaimoNiskanen commented:
@bjorng: I you are right. There should not be any accidantal `handle_call` nor `handle_cast` since they ought to have a corresponding API function that causes them. Therefore they should be optional in the behaviour.  There are odd cases for example if a client uses the server in a different server's API , but that is maybe not worth considering..

@zhird: The code needed to translate `{error,undef,[{Module,Callback,Args},...]}` into `{error,(undef_callback,{Module,Callback,Args}),[...]}` clutters the code for every callback call in `gen_*.erl`, and I am starting to have second thoughts about how much clearer it gets.  Or maybe I am just used to read stacktraces...

I notice that this translation is today only done in `gen_statem.erl`, so the question is if it should be removed therefrom instead...?

zhird commented:
@RaimoNiskanen: Yes it got more cluttered than I thought it would, and after looking at it a bit more I don't think it's worth it. So, I'll keep the `undef` as error reason.

ferd commented:
I tend to agree with @RaimoNiskanen's prior post regarding defaults:

- The callbacks `handle_call/2,3`, `handle_cast/2`, `handle_info/2,3`, `handle_event/3` and `handle_sync_event/4` should all crash the server at runtime if they are not implemented. 
- There is no good default for them. Log vs. no log, crash vs. no-crash, and in the case of sync calls, just eating the message to make the other end die vs. responding
- same vision for other callbacks (`code_change`, `terminate`, etc.)

I'm kind of wondering what would be the point of declaring behaviours if we don't get warnings about incomplete implementations? I guess there's always the type spec bit of it, but what is the feature gonna be if 100% of the callbacks are optional? 

If you want a default where you crash and want to stop getting warnings, just remove the `-behaviour(gen_server).` line or leave it in a comment or something? Then you'll get your super fun "don't warn me if I'm missing stuff and also crash at runtime." I'm not sure what I'm missing here.

RaimoNiskanen commented:
@ferd: Good point!

Assuming that point of view we still can question if there are parts of a these behaviours that is so annoying often enough to implement that they should be optional in the behaviour.  Currently on that list:
- `init` that is annoying to implement the times you do not use it because you use `enter_loop` and feel frustrated having to write a dummy just to get your module through the compiler.
- `terminate` where it is annoying having to write a dummy that returns `ok` almost every time.

essen commented:
I think the arguments about "having to write it when you don't use it" are a bit flawed. Most people use templates. Very few people get frustrated by having to implement callbacks they don't use. People who use templates aren't going to stop using them just because you make the callbacks optional, it's still quicker to go with a template.

The init/enter_loop has never been an issue for me. A module like this uses enter_loop, implements all the callbacks including init, and produces no Dialyzer warnings: https://git.ninenines.eu/ranch.git/tree/examples/tcp_reverse/src/reverse_protocol.erl - perhaps it should, but it doesn't right now.

I also agree that there's no good defaults for most callbacks. There's no consensus at the very least. So making them optional with an arbitrary default does not seem that helpful.

I'm all for making terminate optional NOT because it's annoying to write, but BECAUSE it should almost never be implemented - monitors are better for cleaning up. Having it mandatory makes people want to use it. It should instead be hidden and under lock and key if you ask me.

IngelaAndin commented:
I use terminate, but if I am not mistaking it will only be called if your process traps exits, which might not be obvious. 

I consider all code that is not used to be obfuscation the implementation.

zhird commented:
Yes, good point @ferd! 

I can see the value of getting warned if the behaviour specific functions aren't implemented, but I also agree with @IngelaAndin that you shouldn't have to write dead code! Leaving the behaviour declaration in a comment is one solution for that though... So, now I'm leaning towards that keeping the behaviour specific functions (the `handle_*`ones) mandatory is a better idea.

`terminate` and `code_change` should be optional anyway, since they are so rarely used, but I'm not sure about `init` since I think it's quite rare for it to be skipped, right?

essen commented:
`init` is rarely skipped because people send a message to `self()` instead of using `enter_loop`, and open themselves up to race conditions. Few people use `enter_loop` when they should. And as I pointed out you can very well make `enter_loop` call `init`, so there's really no need for making it optional.

IngelaAndin commented:
@ferd @zhird  I agree that having a behaviour with no mandatory functions feels very strange.  

I think that originally behaviours where thought of only as process behaviours. But that is to far back in history for me to know, so this is my perception.  And over time it was evident that behaviours is a good way to specify APIs in general.  Not all behaviours in OTP are process behaviours, ex ssl_session_cache_API. And with spec added it became possible to have optional callbacks, which does have a use case.

In my experience I usually have use cases for both call, cast and info in  my processes, and if call or cast is not used I will have a crash case for it, and that can be acceptable and not too cluttering. 

I use enter_loop sometimes to avoid the synchronous init when it is not needed and sending yourself init ack will become a bottleneck. But I will have a init function that is used but not by the behaviour, but I think @RaimoNiskanen  has a point that it is a bit confusing that it is mandantory.

IngelaAndin commented:
I however think that having a code_change catch all  (as done in templates, I have them too) is cluttering. If you need a code_change the whole point is that it should do state transformation.  And you only want to have the code_change transformations supported by your corresponding appup. Next time your make a new version of your application you probably only want to support soft upgrade from the previous version and sometimes soft upgrade  is too complicated to be worth it or not even possible.

zhird commented:
I pushed a new version:

I think the `handle_info` functions are just a way to handle the special cases when you get an ordinary banged message, and often you will not use it. Silently dropping messages was a dangerous idea though, so I've added logging through `error_logger` there. So if we choose this version and you really don't want the logging you'll have to implement `handle_info`.

`terminate` and `code_change` are also only used in special cases so they too are optional. `terminate` has the default of doing nothing, but `code_change` crashes since you should not forget it if you add the `Change={advanced, Extra}` in the appup file.

I've also chosen to add `init` as optional since the `enter_loop` function is available.

`gen_server`:
I think `handle_call` and `handle_cast` is what's special about the `gen_server`, so they are kept mandatory while the other ones are optional.

`gen_event`:
Here I find that `handle_event` is the only special function that should be mandatory. `handle_call` is, as I far as I can see, used in special cases only, so I made that one optional too. It defaults to a crash since you should not end up there by mistake.

`gen_statem`:
@RaimoNiskanen already made the changes of `terminate` and `code_change` [here](https://github.com/erlang/otp/commit/0f587ce17f4ad292f6f2d23d6244426046134f38) and [here.](https://github.com/erlang/otp/commit/913d0b52df1e029fb1728b44ba7da318f3dc49dd) So in `gen_statem` all callbacks except `callback_mode` are optional. I just added some tests and updated the documentation a bit.

`gen_fsm`:
`handle_event` and `handle_sync_event` is kept mandatory.

So, what do you think about this?

essen commented:
I don't think `handle_info` is as rare as you make it sound, it's not just the bang (at least from the user's perspective), it's also timers (send_after, start_timer and friends), monitors, links, process groups (global:send and friends) and asynchronous interfaces ({active, once} and friends). Only the simplest behaviors don't end up implementing `handle_info`. That said, if it is to be optional, then logging sounds sensible enough.

RaimoNiskanen commented:
@essen: I agree.  In many work areas most serious servers need to implement `handle_info`.  But for examples and small simple servers it is not needed, and having a default logging implementation forces less boilerplate code, which is probably a good thing.  At least beginners and example writers will appreciate this.

ferd commented:
But does such a default (logging but not crashing) kind of contradict the idea of failing fast on unknown errors?

Like this kind of approach, when not done consciously by the programmer, seems to go against the very basic principles of error handling in Erlang.

I'm still not sure why people who don't want to implement all the callbacks don't just comment out the `-behaviour` line in their code.

zhird commented:
@essen, yes I made that sound a bit too rare, but I still think it's for special cases and not for the behaviours. 

@ferd, dialyzer also checks arguments and returns, and as it would work here you would get warnings about what I find to be behaviour specific functions. And I'm not sure how much value is added by warning about the other functions? I don't find that very valuable. About logging, not crashing; I just don't see crashing as a usable alternative. So in that case I would prefer to have `handle_info` mandatory, which is still an alternative.

okeuday commented:
@zhird I agree that it is safest to have unused callbacks cause a crash to be fail-fast.  Using `error_logger` isn't a good solution due to its throughput limitations, and the potential for its output to be ignored when other logging is used.  For example, if a `gen_server:call` throws a timeout exception that is caught, an info message will still be received and this `handle_info` message may be surprising when the person does not implement `handle_info`, however, it is important to have the process crash rather than risk throughput problems with `error_logger` or other issues the (possibly ignored) failure may lead to.

essen commented:
But a crash results in a crash report (possibly followed by progress reports) so wouldn't it have the same "issue" you mention anyway?

okeuday commented:
@essen Not in the same way, since the crash only occurs once, or possibly a number of times limited by a restart count if a supervisor is being used correctly.  That avoids the output being unbounded.  The `sasl` output often goes to a different file than `error_logger` output, so the two really represent separate information.

josevalim commented:
> I'm kind of wondering what would be the point of declaring behaviours if we don't get warnings about incomplete implementations? I guess there's always the type spec bit of it, but what is the feature gonna be if 100% of the callbacks are optional?

Since Elixir has provided optional callbacks for a while, I would like to add that this is somewhat of a pain point. Someone would accidentally implement `handle_info/3` instead of `handle_info/2` and then wonder why their code is not being invoked. This is less of an issue if the missing implementation leads to an error or logs, as it is being discussed for `handle_info/3`, but it can be annoying on callbacks like `terminate/2`, where a typo in the name or a wrong arity won't emit any error nor print any warning.

fenollp commented:
@josevalim Isn't this then an Elixir-specific request?

josevalim commented:
@fenollp I didn't request anything. :) @ferd, who I quoted, was wondering the point of declaring behaviours since developers will no longer get warnings, and I wanted to add to his concerns that, in our experience, the lack of warnings can be an issue in the scenarios I described. Since callbacks are being made optional in OTP, there is a chance Erlang will face the same issues that we currently do. At the moment, I am not proposing nor requesting any solution to the problem, I just wanted to bring the problem to light.

zhird commented:
@josevalim, thanks for the input! How do you feel about the callbacks being optional in Elixir? There are arguments both for and against this, so it would be nice with some more input from the Elixir community since you've already the opportunity to really try this out.

josevalim commented:
@zhird making callbacks optional in Elixir made it clear that warnings provide guidance on what needs to be implemented. Once the callbacks were made optional, the guidance and compile-time feedback was gone.

Of course this is a tricky line to walk on, because if you have too many warnings, then they just become a nuisance. That's probably why everyone agrees that `format_status` should be an optional callback, especially as it is very rarely implemented. 

Therefore my current state of mind is that a callback should be made optional if 1) it is rarely implemented or 2) we can provide guidance for the cases the callback is not implemented but attempted to be used.

Guidance here can come in the shape of descriptive errors or log messages. For example, if `handle_call/3` is not implemented and someone invokes `gen_server:call/3`, you could error or log by saying:

```
gen_server:call/3 was invoked but there is no handle_call/3 implementation in ?MODULE.

handle_call/3 is a gen_server callback that receives three arguments, the message,
the from tuple and the state. It may return a reply or not. See xyz for more information.
```

The first line is all an advanced developer needs. The remaining of the message will provide further directions for those who need.

This approach can be applied to callbacks such as `handle_call`, `handle_cast` and `code_change`. Folks have raised concerns over raising or logging in `handle_info` though. I am not sure what to make of `init` and `terminate`.

IngelaAndin commented:
When it comes to format_status it was added way later than the other gen functions. I think it is quite useful and fairly anonymous as it is optional.  However I believe there was no grand reason for it becoming optional but rather it was not though of. It feel it could be optional just saying it being so now is not much of an argument.

josevalim commented:
Thanks @IngelaAndin for the correction. Would you also say it was not made required for backwards compatibility reasons? Otherwise all previous gen servers implementations would now break under some circumstances as they didn't have the callback defined?

RaimoNiskanen commented:
@josevalim: I think backwards compatibility was a big reason to make `format_status/2` optional.

IngelaAndin commented:
@josevalim,  @RaimoNiskanen  I agree that backwards compatibility would have been a strong argument for letting it be optional.

nathanaschbacher commented:
I have to say I'm in agreement with @ferd here.  This seems like a solution that turns a write/compile-time annoyance (that could be ameliorated entirely by adding module templates to rebar3 or to your editor) into an irritating potential runtime debugging problem.

Looking across the pond a bit... One of the more frustrating parts of OOP software reliability is when you're getting almost nonsensical, but not entirely absurd, base-class behavior from your application code that's consuming a derived class that doesn't override everything. It's not entirely clear to me what's gained by adding this kind of non-explicit default implementation to Erlang behaviors, but it's clear to me what will be sacrificed to get there.

Why not just add canonical templates to rebar3 akin to `rebar3 new <behavior name> my_server_thingy` that dumps a templated `my_server_thingy.erl` file into your `src` directory?  Then you can trivially edit what the "default" implementation is if you want to.  Modifying how behaviors work in OTP directly doesn't make a lot of sense to me.  This seems like a problem that more complete tooling can fix.

Apologies for terseness and typos, posting from phone.

zhird commented:
We discussed this in the team and think that `handle_info` (default: logging), `terminate` (default: no cleanup) and `code_change` (default: crash) should be optional while the rest should be mandatory.

We think that would make it less "scary" for newcomers to erlang when implementing their first behaviors and make it possible to make code cleaner (for those who care about that), while not causing very noticeable inconveniences (especially not if you use templates).

Thoughts?

essen commented:
Has there been any study of real code that shows that `handle_info` is not used that often to warrant making it optional? Did you make a survey asking newcomers what they actually think about it?

From the start this PR seems to try to address perceived feelings and personal bias more than anything else. You might want to have real data before making a decision.

While it's hard to find code doing something in `code_change`, we don't really know how much the others are implemented. If `handle_info` is implemented 75% of the time in real code it's probably a bad idea to make it optional.

josevalim commented:
@zhird my first thought was not include `handle_cast` as well. In my experience, I use `handle_info` more frequently than `handle_cast`. However, that's my personal bias and I agree with @essen. Measuring usage of those callbacks can help us be more objective.

Maybe the best option is to make `code_change` and `terminate` optional for now, since there seems to be no objection in making those optional and everyone seems to agree on what their default implementation should be. Plus `terminate` requires trapping exits and other pitfalls which would justify making it optional regardless of the newcomer experience.

zhird commented:
Thanks for the input but we don't think a survey is needed. If you think these warnings are that important you are of course welcome to make a one.

I'll merge the changes soon, so they will get into the release candidate.

mrallen1 commented:
For the record - I am +1 on terminate and code_change as optional with sensible defaults. I wrote a whole article about [when OTP decides to call terminate/2](https://gist.github.com/mrallen1/806fe5506132260574af33e99dadd499) which is current behavior as of 18.3.3 if you're interested in that topic.

I am very -1 on making handle_info optional. I can't think of a time when all but the most elementary of gen_server implementations did not need to implement that call back to handle messages from sources which are not internal to the gen_server:call/cast API.  It's a very very common pattern.

I would ask you to please reconsider this decision. Thanks.

dalizard commented:
If it ain't broke, don't fix it. Just saying.

brunogirin commented:
As @nathanaschbacher mentioned looking over the pond at OO languages, the way Java does it is explicit: when you have a complex interface (class where all methods are abstract), the API also has a DefaultXXX or AbstractXXX class that provides a full or partial implementation of that interface with default behaviour. By doing this, the developer can choose to use the default behaviour in cases where they have few things to override or to fully implement the interface when they need custom behaviours. This results in a fair few additional classes but makes all this explicit and gives developers a choice. Would a similar solution work for Erlang?

mrallen1 commented:
Extremely -1 on adding notions of inheritance to Erlang modules.

nathanaschbacher commented:
Likewise.  I intended my comment to indicate a sense of foreboding.

On Apr 26, 2017 12:08, "Mark Allen" <notifications@github.com> wrote:

> Extremely -1 on adding notions of inheritance to Erlang modules.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/erlang/otp/pull/1346#issuecomment-297461003>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAShBXXqQPb_mxItHKYQ-INtx0-MQylNks5rz2wEgaJpZM4MD6e6>
> .
>

brunogirin commented:
Fair enough :-) Irrespective of inheritance, what Java does is give devs the choice: do you want the behaviour with no defaults or one with some defaults? So I was wondering whether it would make sense for Erlang to do something similar and let developers choose whether they want defaults or not. The downside of this is, if the aim is to simplify the development process, this would add one more decision for devs to take, which might be counter productive.

IngelaAndin commented:
@mrallen1  I was not really a fan of making handle_info optional, however the rational of changing the behaviour, so that more  callbacks becomes optional,  was to make the school books examples easier.  This will lower the thresholds for beginners. From that point of view it makes some sense.

essen commented:
This doesn't lower the threshold, they will still need to learn the callback. Besides, most books teach people to write their own gen_server first so this is a moot point.

mrallen1 commented:
@IngelaAndin Thanks for your reply - I appreciate the goal of making Erlang easier for beginners. Indeed, I welcome simplification when it makes sense. 

From my perspective, I think it's important for someone who is starting out to understand how messages which are sent into a gen_server and don't come from the gen_server:call/cast API are handled. 

One of Erlang's great strengths is the design decision that explicit is preferable to implicit - there is no (ok, little 😄) confusing magical behavior - by and large, Erlang does what you see in your editor window. From that viewpoint, it isn't a simplification to move away from that design decision for the handle_info callback.

josevalim commented:
I don't think making it optional makes Erlang (or `gen_server`) more or less explicit. The `gen_server` itself is already a large abstraction that does a lot for developers and I would say an optional callback does not add or remove much compared to that. There are others features in `gen_server` that could be done more explicitly, such as registering names, handling timeouts, but we have accepted them as part of the abstraction or part of the API.

We already had discussions in the mailing list about the complexities from moving from a traditional imperative language to Erlang. Not only we have many functional concepts, we also have the concepts behind fault-tolerance and concurrency. That's why I personally prefer to expose developers to the minimum amount of concepts as they learn. There is no need to talk about `terminate` or `code_change` on your first encounter.

I even think the call vs cast dichotomy is unnecessary. Most developers should be using `call` exclusively. I have already seen over and over again new developers reading the documentation for `call` and `cast` and choosing `cast` because "it is asynchronous and therefore it must be faster". Even with explicit disclaimers that `call` should be preferred as it provides better error semantics and back-pressure.

Those who like to learn by reference or those who like to teach by building a `gen_server` by hand will continue to get and give a full reference. But not everyone learns from books. And not all learning materials prefer this approach. And if multiple resources teach `gen_server` from inside out, maybe it is a consequence that it is really hard to teach a `gen_server` without explaining all of the related concepts and we need to consider if that's good or bad.

In any case, developers on their second rodeo can always get more information from the documentation. Ideally alongside examples of when such callbacks would be necessary, such as using `cast` in case of rare deadlocks or using `info` for handling monitoring messages. I would say good error messages can even do a better job of teaching rather than always generating handle_info. What if the `gen_server` logged this on unknown messages:

> [info] gen_server <0.33.0> has received message ... but it has not implemented a handle_info/2 callback. 
> [debug] The handle_info callback receives the message itself and the current server state. The callback must be implemented whenever the gen_server needs to receive direct messages, such as messages sent with ! or coming from erlang:monitor/2. See x for more information.

I also don't think making it optional hurts when reading code. If I am reading a `gen_server` implementation and I need to find out how it behaves in `handle_info`, I typically go to the function index or search for the function name. The amount of work in finding out the function was not defined or in reading `handle_info(_Info, State) -> {noreply ,State}.` is roughly the same.

PS: I originally wrote that I agreed that `handle_info/2` should not be made optional but I changed my mind as I wrote this reply. 🙄

essen commented:
The complexity for beginners does not come from handle_info, but rather from having both handle_cast and handle_info. It is even more regrettable because while casts could very well be normal messages, normal messages can't always be casts!

A good gen_server interface for beginners should have *call* and *info*, not *call* and *cast*. And that's often what I end up teaching when doing it "from scratch" (which doesn't take all that long and is a great experience because it's an exercise that confirms you understood everything you learned about processes so far). There is no answer to beginners for having separate cast and info that is not completely pedantic to them. Making it optional will not change that. What it will do however, is provide a good WTF when they discover that there is a 3rd callback after all and that it's basically casts except not really, and there won't be any good answer to give to explain why it's hidden by default.

Anyway, can we stop talking about feelings and imaginary people and start looking at real projects and ask real beginners? Do some A/B testing on training classes for example to see which approach works best?

josevalim commented:
> A good gen_server interface for beginners should have call and info, not call and cast. 

I agree `cast` should be avoided as much as possible. Why not `handle_cast` was my first thought after `handle_info` was announced as optional.

> Anyway, can we stop talking about feelings and imaginary people and start looking at real projects and ask real beginners?

FWIW, my comments above have been drawn from my experience in projects, trainings and in helping the community. I have been able to compare and discuss Erlang and Elixir approaches multiple times, which are polar opposites on this topic. While there is likely a bias, it is far away from being a feeling or imaginary.

essen commented:
We all draw from experiences, but only have a small view of the world, and are certainly not beginners. We need beginners to understand what is best for beginners. We can't just change things because we feel beginners would have it easier, without testing these changes with beginners directly.

nathanaschbacher commented:
I'm still not clear why this whole thing is best solved by modifying the
implementation and semantics of things in OTP directly. I must be missing
some context for why that's the ideal organization of concerns.

If the pain point is that it's onerous to implement these common things by
hand, then why can't rebar3 and a set of simple additional tasks and
templates for a "default" gen_server module be canonicalized and used to
alleviate that pain.

That also makes it pretty trivial to adapt the semantics to each person's
desired default by just editing the template.

On Apr 27, 2017 11:01, "José Valim" <notifications@github.com> wrote:

> A good gen_server interface for beginners should have call and info, not
> call and cast.
>
> I agree cast should be avoided as much as possible. Why not handle_cast
> was my first thought after handle_info was announced as optional.
>
> Anyway, can we stop talking about feelings and imaginary people and start
> looking at real projects and ask real beginners?
>
> FWIW, my comments above have been drawn from my experience in projects,
> trainings and in helping the community. I have been able to compare and
> discuss Erlang and Elixir approaches multiple times, which are polar
> opposites on this topic. While there is likely a bias, it is far away from
> being a feeling or imaginary.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/erlang/otp/pull/1346#issuecomment-297740196>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAShBQYmneE2xvRHnhlVPGmT5tU45jtVks5r0K3IgaJpZM4MD6e6>
> .
>

ferd commented:
My point of view is that we probably just traded off "what is handle_info for?" for "why am I getting a log line about handle_info?"

I can't imagine it's any significant and I'll personally keep treating handle_info as mandatory in tutorials and actual code.

IngelaAndin commented:
Well I am not going to advocate for that point as I was not the one making it. @garazdawi  do you have something to add (as I remember it, you where one of the people making this argument)? Or someone else favouring this point of view?

josevalim commented:
> without testing these changes with beginners directly.

I have been doing that, for this particular issue, for almost 3 years. Basically [since this exact feature was added to Elixir](https://github.com/josevalim/xgen) but with larger scope. We had plenty of time to collect feedback from newcomers on this topic. Some of the issues the Elixir community is struggling based on such changes have also been reported at the beginning of this discussion.

@rvirding and @michalslaski may also have experiences to share, as I believe they have given trainings in Erlang and Elixir (which already has this change). They may also be less biased towards optional callbacks than me.

> My point of view is that we probably just traded off "what is handle_info for?" for "why am I getting a log line about handle_info?"

@ferd would you still hold this point of view if we improve the amount of debug information logged when it happens? As proposed [here](https://github.com/erlang/otp/pull/1346#issuecomment-297731437).

ferd commented:
@josevalim we just deferred a compile-time warning to a run-time warning.

ferd commented:
To elaborate a bit, I have no opposition to making `terminate/2` and `code_change/3` optional or have empty defaults since they're usually easy to figure out.

There are however like 3 levels here to the `handle_info` default:

1. presume there is no sane default, and make no decision. Additionally, warn at compile time that things will fail at run-time.
2. presume it is a safe default to discard. Additionally, warn at compile time that you should patch the code if you don't like this behaviour
3. presume it is a safe default to discard. Additionally, warn at run-time so that your users can tell you that you should patch your code if they don't like that behaviour

People reading this thread somehow felt that option 3 would be the most appropriate one. I can't understand that the following were all true at once:

- it was decided there is an appropriate default behaviour
- it was decided that the developer should not receive a warning about it anymore
- it was decided that the user of the code should be told about it at run time rather than the developer

You could reasonably argue for a default behaviour (I don't really agree, but can concede the point). You could also argue that no warning is acceptable (I don't really agree either, but can begrudgingly concede if you instead mention it a lot in the docs and then point people to it). You could maybe warn both at compile-time and run-time (which I don't 100% agree but can see the logic fine).

You can't reasonably argue that it's preferable to warn at run-time than compile time though. You just can't.

Program users can't do anything about logs intended for the developer, and the developer (who I expect must compile their code to run it) would benefit from the warning if we consider things to be warning-worthy.

Right now we consider things to be warning-worthy but have decided to defer it to run time for a reason I can't understand. I don't want to assume "but it looks so much cleaner when it compiles without warning" has been judged acceptable because what the hell if that is so it would mean we prefer to have a clean build log to a clean program run and what am I even doing programming in a compiled language if that's so.

josevalim commented:
Thanks @ferd, that's a very good analysis. I would like to add two points:

  1. I believe 3 was chosen because there is an assumption warnings are not going to be triggered unless there are stray messages, which you will hopefully detect during development and testing. Otherwise you will have to fix once you notice them in production. So it is an exchange for an always present warning in favor of a warning that is only emitted when you need to act on it.

  2. A lot of people use templates today to get rid of the boilerplate or copy a default implementation. There is a high chance that this default is simply: `handle_info(_Msg, State) -> {noreply, State}.`. Is discarding all messages a better default than also logging when they are discarded?

ferd commented:
For point 1. I'd say then that the logical thing might be to warn both at compile time and test time at the very least.

For point 2. I'd say it's essentially better because at least there is some visible source code under your own repository for the behaviour you see as a default.

okeuday commented:
@josevalim As described above (https://github.com/erlang/otp/pull/1346#issuecomment-285147990) we should have optional callbacks crash if they aren't declared, to keep usage safe.  Otherwise, we would encouraging silent messaging problems or logging throughput problems.

rvirding commented:
I agree with Fred here, especially with handle_info. It is a critical behaviour so not having defined should lead to a crash, not just a warning.

Not defining callbacks can lead to some very strange behaviour. While it might save a few lines when writing the module (very few actually when you think about it) it makes it much more difficult to understand other peoples code, which is critical for more long lived projects. I have experience of this and not knowing whether the callbacks were there would have made it extremely difficult to understand the code.

So don't get rid of warnings and if handle_info is to be optional then the default should crash the server.

vans163 commented:
Just so your aware this breaks Elixirs recompiler.  I really wish this behavior can be behind a commandline flag with the default on.

When you call recompile in Elixir, it fully unloads modules then brings them back.  If you edit the code twice to the same module and call recompile twice, it fully purges and unloads plus brings back the module.  This results in any pending send_after, or messages in the mailbox, to be delivered to the purged and unloaded module, the handle_info on that module is non-existant, so it gives a WARNING handle_info is not defined (even thought it is defined in source code, module is just unloaded).  The message gets dropped, before this would result in a crash.

Now you get into a unknown state, as the message got "delivered" but the module wasnt there as its purged+unloaded +loaded, so instead of a crash of the gen_server, to allow the gen_server to rebuild its state. You got a discarded message.
